src/webgpu/shader/execution/expression/call/builtin/textureStore.spec.ts

```diff
@@ -175,91 +103,129 @@ g.test('texel_formats')
     }
   })
   .fn(t => {
-    const { format, _shaderType } = t.params;
+    const { format, stage, access, viewDimension, _shaderType } = t.params;
     const values = inputArray(format);
 
-    let numChannels = 4;
-    switch (format) {
-      case 'r32uint':
-      case 'r32sint':
-      case 'r32float':
-        numChannels = 1;
-        break;
-      case 'rg32uint':
-      case 'rg32sint':
-      case 'rg32float':
-        numChannels = 2;
-        break;
-      default:
-        break;
-    }
-
-    let zeroVal = ``;
-    if (numChannels > 1) {
-      zeroVal = `val[idx % ${numChannels}] = 0;`;
-    }
-
-    let wgsl = `
-const range = array(`;
-    for (const v of values) {
-      wgsl += `${v},\n`;
-    }
-
-    wgsl += `
-);
+    const suffix = format.endsWith('sint') ? 'i' : format.endsWith('uint') ? 'u' : 'f';
+    const swizzleWGSL = viewDimension === '1d' ? 'x' : viewDimension === '3d' ? 'xyz' : 'xy';
+    const layerWGSL = viewDimension === '2d-array' ? ', gid.z' : '';
+    const wgsl = `
+const range = array(${values.map(v => `${v}${suffix}`).join(',')});
 
 @group(0) @binding(0)
-var tex : texture_storage_1d<${format}, write>;
+var tex : texture_storage_${viewDimension.replace('-', '_')}<${format}, ${access}>;
+
+fn setValue(gid: vec3u) {
+  let ndx = gid.x + gid.y + gid.z;
+  let vecVal = vec4(
+    range[(ndx + 0) % ${values.length}],
+    range[(ndx + 1) % ${values.length}],
+    range[(ndx + 2) % ${values.length}],
+    range[(ndx + 3) % ${values.length}],
+  );
+  var val = vec4<${_shaderType}>(vecVal);
+  let coord = gid.${swizzleWGSL};
+  textureStore(tex, coord${layerWGSL}, val);
+}
 
 @compute @workgroup_size(${values.length})
-fn main(@builtin(global_invocation_id) gid : vec3u) {
-  let idx = gid.x;
-  let scalarVal = range[idx];
-  let vecVal = vec4(scalarVal);
-  var val = vec4<${_shaderType}>(vecVal);
-  ${zeroVal}
-  textureStore(tex, gid.x, val);
+fn cs(@builtin(global_invocation_id) gid : vec3u) {
+  setValue(gid);
+}
+
+struct VOut {
+  @builtin(position) pos: vec4f,
+  @location(0) @interpolate(flat, either) z: u32,
+}
+@vertex fn vs(
+  @builtin(vertex_index) vNdx: u32,
+  @builtin(instance_index) iNdx: u32,
+) -> VOut {
+  let pos = array(vec2f(-1, 3), vec2f(3, -1), vec2f(-1, -1));
+  return VOut(vec4f(pos[vNdx], 0, 1), iNdx);
+}
+
+@fragment fn fs(v: VOut) -> @location(0) vec4f {
+  setValue(vec3u(u32(v.pos.x), u32(v.pos.y), v.z));
```

Comment:
v.pos.xy is in -1, 1 range. I think you want is uint coord here (something like gl_FragCoord from 0 to tex wdith/height)?

Reply:
v.pos.xy is in texel coordinates (it is `gl_FragCoord`)

```
   0.5, 0.5    1.5, 0.5    2.5, 0.5
   0.5, 1.5    1.5, 1.5    2.5, 1.5
   0.5, 2.5    1.5, 2.5    2.5, 2.5
```

Reply:
It's confusing because 

```
struct VOut {
  @builtin(position) pos: vec4f;
  ...
};
```

In the vertex shader this is normalized coord used to computing triangle vertices.
In the fragment shader it's `gl_FragCoord`.